sub-packages/bionemo-llm/src/bionemo/llm/data/datamodule.py

```diff
@@ -0,0 +1,69 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: LicenseRef-Apache2
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import Any, Dict
+
+import pytorch_lightning as pl
+from nemo.utils import logging
+
+
+class MegatronDatamodule(pl.LightningDataModule):
+    """A mixin that adds a `state_dict` and `load_state_dict` method for datamodule training resumption in NeMo."""
+
+    def __post_init__(self):
+        """Set init_global_step to 0 for datamodule resumption."""
+        self.init_global_step = 0
+
+    def update_init_global_step(self):
+        """Please always call this when you get a new dataloader... if you forget, your resumption will not work."""
```

Comment:
is there no way we can automate this call?

Reply:
I tried moving it to a few different places, it seems like it needs to happen after `setup`. I'm guessing the correct fix is to add a new megatron callback hook at the right point of initialization and have this automatically called on the `trainer.datamodule` and require the datamodule type to support this, or if it doesn't at _least_ log a warning. 

This is SUPER HACKY as is now, and should be handled upstream by NeMo. @marcromeyn @ashors1.

Reply:
So to be clear I think the megatron strategy should prefer to work with a couple key objects and warn if they are not encountered:

1. The DataloaderWrapper thing that communicates the `mode`. There should be a warning if the mode is not found with an explanation of how resumption is either going to go wrong on the trainer or val steps.
2. The DataModule piece needing state, and this global step communication from the trainer. The strategy should warn if a vanilla datamodule is encountered, and use some special datamodule type that exposes these key features for resumption.

Reply:
I am still unfamiliar with its mechanism. Probably @jstjohn can answer more.

Reply:
Should be something I have missed.

Reply:
Sure, so if you forget to update `self.init_global_step` _and_ `self.data_sampler.init_global_step` then the `consumed_samples` calculations are way off. They get inflated and a bunch of samples are skipped on resumption.

Reply:
It's ok to call multiple times apparently.

Reply:
maybe we can try moving this into `def train_dataloader` or something.

Reply:
unless it's required to be called in `def val_dataloader`. I'm not sure exactly where this needs to be called to be honest... Just that it needs to happen after `setup` at some point.